vdr/src/client/client.rs

```diff
@@ -302,64 +322,265 @@ pub mod test {
+        let signature = signer.sign(&sign_bytes, TRUSTEE_ACC.as_ref()).unwrap();
+        transaction.set_signature(signature);
 
-        async fn get_block(&self, _block: Option<u64>) -> VdrResult<BlockDetails> {
-            todo!()
-        }
+        transaction
+    }
 
-        async fn get_transaction(&self, _hash: &[u8]) -> VdrResult<Option<Transaction>> {
-            todo!()
+    pub fn read_transaction() -> Transaction {
+        Transaction {
+            type_: TransactionType::Read,
+            from: None,
+            to: VALIDATOR_CONTROL_ADDRESS.clone(),
+            nonce: None,
+            chain_id: CHAIN_ID,
+            data: vec![],
+            signature: RwLock::new(None),
+            hash: None,
         }
     }
 
-    pub fn mock_client() -> LedgerClient {
-        let mut client = LedgerClient::new(
+    pub fn mock_custom_client(client: Box<dyn Client>) -> LedgerClient {
+        let mut ledger_client = LedgerClient::new(
             CHAIN_ID,
             RPC_NODE_ADDRESS,
             &contracts(),
             Some(&QuorumConfig::default()),
         )
         .unwrap();
-        client.client = Box::new(MockClient {});
-        client
+
+        ledger_client.client = client;
+        ledger_client
     }
 
     mod create {
+        use mockall::predicate::eq;
+        use serde_json::Value;
+
         use super::*;
 
         #[test]
         fn create_client_test() {
             client();
         }
+
+        #[test]
+        fn create_client_invalid_contract_data() {
+            let contract_config = vec![ContractConfig {
+                address: VALIDATOR_CONTROL_ADDRESS.to_string(),
+                spec_path: None,
+                spec: Some(ContractSpec {
+                    name: CONTRACT_NAME_EXAMPLE.to_string(),
+                    abi: Value::String("".to_string()),
+                }),
+            }];
+
+            let client_err = LedgerClient::new(CHAIN_ID, RPC_NODE_ADDRESS, &contract_config, None)
+                .err()
+                .unwrap();
+
+            assert!(matches!(
+                client_err,  | VdrError::ContractInvalidInputData { .. }
+            ));
+        }
+
+        #[test]
+        fn create_client_invalid_node_address() {
+            let client_err = LedgerClient::new(CHAIN_ID, "..", &contracts(), None)
+                .err()
+                .unwrap();
+
+            assert!(matches!(
+                client_err,  | VdrError::ClientNodeUnreachable { .. }
+            ));
+        }
+
+        #[test]
+        fn create_client_contract_path_and_spec_provided() {
+            let contract_config = vec![ContractConfig {
+                address: VALIDATOR_CONTROL_ADDRESS.to_string(),
+                spec_path: Some(build_contract_path(VALIDATOR_CONTROL_PATH)),
+                spec: Some(ContractSpec {
+                    name: CONTRACT_NAME_EXAMPLE.to_string(),
+                    abi: Value::Array(vec![]),
```

Comment:
do we have a test for case when spec does not exist by the specified path?

Reply:
Added test